pkg/oc/cli/cmd/image/mirror/mirror.go

```diff
@@ -0,0 +1,667 @@
+
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/kubernetes/pkg/kubectl/cmd/templates"
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+
+	imageapi "github.com/openshift/origin/pkg/image/apis/image"
+	"github.com/openshift/origin/pkg/image/importer"
+)
+
+var (
+	mirrorDesc = templates.LongDesc(`
+		Mirror images from one image repository to another.
+
+		Accepts a list of arguments defining source images that should be pushed to the provided
+		destination image tag. The images are streamed from registry to registry without being stored
+		locally. The default docker credentials are used for authenticating to the registries.
+
+		When using S3 mirroring the region and bucket must be the first two segments after the host.
+		Mirroring will create the necessary metadata so that images can be pulled via tag or digest,
+		but listing manifests and tags will not be possible. You may also specify one or more
+		--s3-source-bucket parameters (as <bucket>/<path>) to designate buckets to look in to find
+		blobs (instead of uploading). The source bucket also supports the suffix "/[store]", which 
+		will transform blob identifiers into the form the Docker registry uses on disk, allowing
+		you to mirror directly from an existing S3 backed Docker registry.
+
+		Experimental: This command is under active development and may change without notice.`)
+
+	mirrorExample = templates.Examples(`
+# Copy image to another tag
+%[1]s myregistry.com/myimage:latest myregistry.com/myimage:stable
+
+# Copy image to another registry
+%[1]s myregistry.com/myimage:latest docker.io/myrepository/myimage:stable
+
+# Copy image to S3 (pull from <bucket>.s3.amazonaws.com/image:latest)
+%[1]s myregistry.com/myimage:latest s3://s3.amazonaws.com/<region>/<bucket>/image:latest
+
+# Copy image to S3 without setting a tag (pull via @<digest>)
+%[1]s myregistry.com/myimage:latest s3://s3.amazonaws.com/<region>/<bucket>/image
+
+# Copy image to multiple locations
+%[1]s myregistry.com/myimage:latest docker.io/myrepository/myimage:{stable,dev}
+
+# Copy multiple images
+%[1]s myregistry.com/myimage:latest=myregistry.com/other:test \ 
+  myregistry.com/myimage:new=myregistry.com/other:target
+`)
+)
+
+type DestinationType string
+
+var (
+	DestinationRegistry DestinationType = "docker"
+	DestinationS3       DestinationType = "s3"
+)
+
+type Mapping struct {
+	Source      imageapi.DockerImageReference
+	Destination imageapi.DockerImageReference
+	Type        DestinationType
+}
+
+type pushOptions struct {
+	Out, ErrOut io.Writer
+
+	Mappings []Mapping
+	OSFilter *regexp.Regexp
+
+	FilterByOS string
+
+	Insecure  bool
+	SkipMount bool
+	Force     bool
+
+	AttemptS3BucketCopy []string
+}
+
+// NewCommandMirrorImage copies images from one location to another.
+func NewCmdMirrorImage(name string, out, errOut io.Writer) *cobra.Command {
+	o := &pushOptions{}
+
+	cmd := &cobra.Command{
+		Use:     "mirror SRC DST [DST ...]",
```

Comment:
Why don't use `skopeo copy` for this?

Reply:
skopeo copy today is optimized for single use cases.  This is a start at a bulk copy, with a particular focus on two areas:

1. being able to take structured input (not just a list, but potentially tens or hundreds of images)
2. being able to run and retry individual operations and display progress

so that we can run this as a job on the user's behalf to mirror image streams from openshift to another registry.  I think that eventually this may converge with skopeo, but I want to iterate towards that use case.  We would probably run this in two contexts:

1. admin running this locally to copy 10-30 images in bulk to another location (as we often do on release push)
2. being a controller driven job, where the controller launches a pod that accepts JSON and returns JSON for structured output

That's just different enough from skopeo that I don't want to start by hacking up copy, but instead come back around to it later.

Reply:
So I'm understanding correctly, today skopeo cannot take multiple inputs for the copy operation and isn't easily run as a job in a pod.

The multiple inputs seems fairly straightforward, though I would say that a wrapper to just run the command multiple times is a good first step.   What is the catch for running in a pod, the output parsing?  Retry seems like a function of job failure (especially if each job were a single copy)

Reply:
The {} syntax was actually bash.

The output needs to be some structured form that a caller can interpret, like JSON.  Retry behavior, parallelism, throttling, caching, and avoiding duplicate copy are also extra work in bash.  Filtering manifest lists and the S3 behavior are ones that would have no analogue in skopeo today but could be relevant in the future.

Reply:
ah.  then i don't think the {} should be represented in the help since it isn't cross-platform.

Reply:
If we represent pipe or redirect in the help anywhere it's the same problem.